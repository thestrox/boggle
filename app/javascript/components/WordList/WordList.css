.wordlist {
    border: 1px solid lightgray;
    padding-top: 0 !important;
    padding-bottom: 0 !important;
    overflow: auto;
}

.wordlist-item {
    justify-content: space-between !important;
}