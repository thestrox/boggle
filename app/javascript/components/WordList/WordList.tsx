import React from "react";

export type WordListProp = {
    wordScoreMap: { [key: string]: number }
}

export default function WordList({ wordScoreMap }: WordListProp) {

    return (
        <ul>
            {Object.keys(wordScoreMap).map(word => {
                return (
                    <li key={word}>{word} : {wordScoreMap[word]}</li>
                );
            })}
        </ul>
    )
}