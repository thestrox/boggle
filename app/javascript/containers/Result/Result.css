.scorelist-container {
    min-width: 200px;
    min-height: 200px;
    max-height: 500px;
}

.total-container, .time-container {
    display: flex;
    padding: 8px;
    justify-content: space-between;
}