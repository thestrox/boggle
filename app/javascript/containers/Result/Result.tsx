import React from "react";
import { connect } from "react-redux";
import { AppState } from "../../store";
import WordList from "../../components/WordList/WordList";
import "./Result.css"
import { Button } from "@material-ui/core";
import { formatTimeDelta, calcTimeDelta, zeroPad } from "react-countdown";
import { Redirect } from "react-router-dom";
import { reset } from "../../store/common-actions";

export type ResultProps = {
    wordScoreMap: { [key: string]: number };
    gameStartDate: number;
    resetBoard: () => void;
}

export type ResultState = {
    toHome: boolean;
}

class Result extends React.Component<ResultProps, ResultState> {
    constructor(props: ResultProps) {
        super(props);
        this.state = {
            toHome: this.props.gameStartDate ? false: true
        };

    }

    toHome = () => {
        this.props.resetBoard();
        this.setState({ toHome: true })
    }

    render() {
        if (this.state.toHome) {
            return (
                <Redirect to='' />
            );
        }

        const values = Object.values(this.props.wordScoreMap)
        const totalScore = values.length > 0 
        ? values.reduce((p, c) => p + c) 
        : 0;

        const formatted = formatTimeDelta(calcTimeDelta(Date.now() - this.props.gameStartDate, { controlled: true }))

        return (
            <div className="result-container">
                <span>Your Result</span>
                <WordList wordScoreMap={this.props.wordScoreMap} className="scorelist-container"></WordList>
                <div className="total-container">
                    <span>Total Score:</span>
                    <span>{totalScore}</span>
                </div>
                <div className="time-container">
                    <span>Time Elapsed:</span>
                    <span>{zeroPad(formatted.minutes)}:{zeroPad(formatted.seconds)}</span>
                </div>
                <Button onClick={this.toHome} variant="contained" color="primary">Home</Button>

            </div>
        );
    }
}

const mapStateToProps = (state: AppState) => {
    return {
        wordScoreMap: state.score.wordScoreMap,
        gameStartDate: state.board.gameStartDate
    }
}

const ResultConnected = connect(
    mapStateToProps,
    {resetBoard: reset}
)(Result)
export default ResultConnected
