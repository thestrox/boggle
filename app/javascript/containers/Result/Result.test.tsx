import React from 'react';
import { shallow } from 'enzyme';
import {Result} from "./Result"
import WordList from '../../components/WordList/WordList';
import { Button } from '@material-ui/core';

function setup() {
    const props = {
        wordScoreMap: {ONE: 3, TWO: 3},
        gameStartDate: 1234,
        resetBoard: jest.fn()
    };

    const component = shallow<Result>(<Result {...props}/>);
    return {
        props,
        component
    }
}

describe('Result Component', () => {
    it('should render correctly with given props', () => {
        spyOn(Date, "now").and.returnValue(1234);
        const {component, props} = setup();
        expect(component.find('div').at(0).hasClass('result-container')).toBe(true);
        expect(component.find('span').at(0).text()).toBe('Your Result');
        expect(component.find(WordList).prop('wordScoreMap')).toBe(props.wordScoreMap);
        expect(component.find('span').at(2).text()).toBe("6");
        expect(component.find('span').at(4).text()).toBe('00:00');
    });

    it('should call props resetBoard on Home button click and change state.toHome to true', () => {
        const {component, props} = setup();
        const instance = component.instance();
        expect(instance.state.toHome).toBeFalsy();
        expect(component.find(Button).simulate('click'));
        expect(props.resetBoard).toHaveBeenCalled(); 
        expect(instance.state.toHome).toBeTruthy();
    });

    it('should have the initial state.toHome to true when gameStartDate prop is null', () => {
        const props = {
            wordScoreMap: {ONE: 3, TWO: 3},
            gameStartDate: null,
            resetBoard: jest.fn()
        };
        const component = shallow<Result>(<Result {...props}/>);
        expect(component.instance().state.toHome).toBeTruthy();

    });

    it('should test snapshot renders', () => {
        spyOn(Date, "now").and.returnValue(1234);
        const {component} = setup();
        expect(component).toMatchSnapshot();
    });
})