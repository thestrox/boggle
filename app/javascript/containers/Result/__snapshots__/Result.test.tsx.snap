// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Component should test snapshot renders 1`] = `
<div
  className="result-container"
>
  <span>
    Your Result
  </span>
  <WordList
    className="scorelist-container"
    wordScoreMap={
      Object {
        "ONE": 3,
        "TWO": 3,
      }
    }
  />
  <div
    className="total-container"
  >
    <span>
      Total Score:
    </span>
    <span>
      6
    </span>
  </div>
  <div
    className="time-container"
  >
    <span>
      Time Elapsed:
    </span>
    <span>
      00
      :
      00
    </span>
  </div>
  <WithStyles(ForwardRef(Button))
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    Home
  </WithStyles(ForwardRef(Button))>
</div>
`;
